pkg/clients/secretsmanager/secret.go

```diff
@@ -0,0 +1,267 @@
+
+// GenerateCreateSecretsmanagerInput from ClusterParameters.
+func GenerateCreateSecretsmanagerInput(name string, p *v1alpha1.SecretParameters, secret []byte) *secretsmanager.CreateSecretInput {
+	c := &secretsmanager.CreateSecretInput{
+		Name:         awsclients.String(name),
+		SecretBinary: secret,
+	}
+
+	if p.Description != nil {
+		c.Description = p.Description
+	}
+
+	if p.KmsKeyID != nil {
+		c.KmsKeyId = p.KmsKeyID
+	}
+
+	if len(p.Tags) != 0 {
+		c.Tags = make([]secretsmanager.Tag, len(p.Tags))
+		for i, val := range p.Tags {
+			c.Tags[i] = secretsmanager.Tag{
+				Key:   aws.String(val.Key),
+				Value: aws.String(val.Value),
+			}
+		}
+	}
+
+	return c
+}
+
+// GenerateUpdateSecretInput from SecretParameters
+func GenerateUpdateSecretInput(name string, p v1alpha1.SecretParameters, secret []byte) *secretsmanager.UpdateSecretInput {
+	u := &secretsmanager.UpdateSecretInput{
+		SecretBinary: secret,
+		SecretId:     awsclients.String(name),
+	}
+
+	if p.Description != nil {
+		u.Description = p.Description
+	}
+
+	if p.KmsKeyID != nil {
+		u.KmsKeyId = p.KmsKeyID
+	}
+
+	return u
+}
+
+// GenerateDeleteSecretInput from SecretParameters
+func GenerateDeleteSecretInput(name string, p v1alpha1.SecretParameters) *secretsmanager.DeleteSecretInput {
+	d := &secretsmanager.DeleteSecretInput{
+		SecretId: awsclients.String(name),
+	}
+
+	switch r := GetRecoveryWindowInDays(p.RecoveryWindowInDays); r {
+	case 0:
+		d.ForceDeleteWithoutRecovery = boolToPtr(true)
+	default:
+		d.RecoveryWindowInDays = IntToPtr(int64(r))
+	}
+
+	return d
+}
+
+// UpdateObservation updates status.AtProvider
+func UpdateObservation(o *v1alpha1.SecretObservation, svr *secretsmanager.GetSecretValueResponse, do *secretsmanager.DescribeSecretOutput) {
+	if svr != nil {
+		svo := svr.GetSecretValueOutput
+		if svo != nil {
+			arn := svo.ARN
+			if svo.ARN == nil {
+				arn = do.ARN
+			}
+			o.ARN = awsclients.StringValue(arn)
+			o.VersionID = awsclients.StringValue(svo.VersionId)
+			o.VersionStages = svo.VersionStages
+
+			if svo.CreatedDate != nil {
+				o.CreatedDate = &metav1.Time{Time: *svo.CreatedDate}
+			}
+		}
+	}
+
+	if do != nil && do.DeletedDate != nil {
+		o.DeletedDate = &metav1.Time{Time: *do.DeletedDate}
+	}
+}
+
+// LateInitialize fills the empty fields in *v1alpha1.SecretParameters with the
+// values seen in secretsmanager.DescribeSecretOutput.
+func LateInitialize(in *v1alpha1.SecretParameters, so *secretsmanager.DescribeSecretOutput) { // nolint:gocyclo
+	if so == nil {
+		return
+	}
+	if in.Description == nil && so.Description != nil {
+		in.Description = so.Description
+	}
+	if in.KmsKeyID == nil && so.KmsKeyId != nil {
+		in.KmsKeyID = so.KmsKeyId
+	}
+
+	r := int64(GetRecoveryWindowInDays(in.RecoveryWindowInDays))
+	in.RecoveryWindowInDays = &r
+
+	if len(in.Tags) == 0 && len(so.Tags) > 0 {
+		in.Tags = make([]v1alpha1.Tag, len(so.Tags))
+		for i, t := range so.Tags {
+			in.Tags[i] = v1alpha1.Tag{
+				Key:   *t.Key,
+				Value: *t.Value,
```

Comment:
```suggestion
				Key:   awsclients.StringValue(t.Key),
				Value: awsclients.StringValue(t.Value),
```

nitpick: we have these small helper functions for these cases. It's not really a big deal here but for example if you were to do operations on pointers in the for loop, then it could add wrong element to the array repeatedly since these `i` and `t` pointers are static and only their values are changing. But helper functions operate in separate stack, they don't have this problem.